### PR TITLE
Parquet Dictionary Predicate Pushdown Fixes

### PR DESCRIPTION
**Pull request by pettyjamesm**
Parquet dictionary pushdown was refactored in https://github.com/prestodb/presto/pull/6892 to remove a nested loop iteration but accidentally left the inner loop `break` statement behind. This meant that dictionary predicate pushdown would read at most 1 dictionary.

In addition to fixing the pushdown behavior, this PR adds support for checking the dictionary pushdown on each column skipping additional dictionary reads once the block can already be filtered.

### CODE REVIEW

File: presto-parquet/src/main/java/io/prestosql/parquet/predicate/PredicateUtils.java [hunk 1]
```diff
@@ -111,22 +110,26 @@ public static boolean predicateMatches(Predicate parquetPredicate, BlockMetaData
...
+                    if (!parquetPredicate.matches(dictionary)) {
+                        return false;
+                    }
+                    dictionaries.put(descriptor, dictionary);
                 }
             }
         }
-        return dictionaries.build();
+        // Attempt full domain check
+        return parquetPredicate.matches(dictionaries.build());
```
Comment (findepi) [inline presto-parquet/src/main/java/io/prestosql/parquet/predicate/PredicateUtils.java]: can this return `true` when each earlier attempt returned `false`?

add explanatory code comment, otherwise it's tempting to replace this line with `return true`

Comment (pettyjamesm) [reply presto-parquet/src/main/java/io/prestosql/parquet/predicate/PredicateUtils.java]: Good point, removed the redundant check at the end.

File: presto-parquet/src/main/java/io/prestosql/parquet/predicate/PredicateUtils.java [hunk 2]
```diff
@@ -111,22 +110,22 @@ public static boolean predicateMatches(Predicate parquetPredicate, BlockMetaData
...
             RichColumnDescriptor descriptor = descriptorsByPath.get(Arrays.asList(columnMetaData.getPath().toArray()));
             if (descriptor != null) {
                 if (isOnlyDictionaryEncodingPages(columnMetaData) && isColumnPredicate(descriptor, parquetTupleDomain)) {
-                    int totalSize = toIntExact(columnMetaData.getTotalSize());
-                    byte[] buffer = new byte[totalSize];
+                    byte[] buffer = new byte[toIntExact(columnMetaData.getTotalSize())];
                     dataSource.readFully(columnMetaData.getStartingPos(), buffer);
-                    Optional<DictionaryPage> dictionaryPage = readDictionaryPage(buffer, columnMetaData.getCodec());
-                    dictionaries.put(descriptor, new DictionaryDescriptor(descriptor, dictionaryPage));
+                    //  Early abort, predicate already filters block so no more dictionaries need be read
```
Comment (findepi) [inline presto-parquet/src/main/java/io/prestosql/parquet/predicate/PredicateUtils.java]: "Early" is no longer as clear as it was, since there is no "normal"

File: presto-parquet/src/main/java/io/prestosql/parquet/predicate/PredicateUtils.java [hunk 3]
```diff
@@ -111,22 +110,22 @@ public static boolean predicateMatches(Predicate parquetPredicate, BlockMetaData
...
+    private static boolean dictionaryPredicatesMatch(Predicate parquetPredicate, BlockMetaData blockMetadata, ParquetDataSource dataSource, Map<List<String>, RichColumnDescriptor> descriptorsByPath, TupleDomain<ColumnDescriptor> parquetTupleDomain)
     {
-        ImmutableMap.Builder<ColumnDescriptor, DictionaryDescriptor> dictionaries = ImmutableMap.builder();
         for (ColumnChunkMetaData columnMetaData : blockMetadata.getColumns()) {
             RichColumnDescriptor descriptor = descriptorsByPath.get(Arrays.asList(columnMetaData.getPath().toArray()));
             if (descriptor != null) {
                 if (isOnlyDictionaryEncodingPages(columnMetaData) && isColumnPredicate(descriptor, parquetTupleDomain)) {
-                    int totalSize = toIntExact(columnMetaData.getTotalSize());
-                    byte[] buffer = new byte[totalSize];
+                    byte[] buffer = new byte[toIntExact(columnMetaData.getTotalSize())];
```
Comment (findepi) [inline presto-parquet/src/main/java/io/prestosql/parquet/predicate/PredicateUtils.java]: nit: ideally code cleanup should go in separate commit

File: presto-parquet/src/main/java/io/prestosql/parquet/predicate/PredicateUtils.java [hunk 4]
```diff
@@ -155,7 +154,7 @@ public static boolean predicateMatches(Predicate parquetPredicate, BlockMetaData
     private static boolean isColumnPredicate(ColumnDescriptor columnDescriptor, TupleDomain<ColumnDescriptor> parquetTupleDomain)
     {
         verify(parquetTupleDomain.getDomains().isPresent(), "parquetTupleDomain is empty");
-        return parquetTupleDomain.getDomains().get().keySet().contains(columnDescriptor);
+        return parquetTupleDomain.getDomains().get().containsKey(columnDescriptor);
```
Comment (findepi) [inline presto-parquet/src/main/java/io/prestosql/parquet/predicate/PredicateUtils.java]: nit: ideally code cleanup should go in separate commit

**Comment by pettyjamesm** [comment]
Test failure is from a stuck container, don’t think it’s related to the changes but I haven’t had a chance to look harder yet.
